pkg/admission/mutator/shoot.go

```diff
@@ -53,21 +67,183 @@ func (s *shootMutator) Mutate(_ context.Context, new, old client.Object) error {
+	s.isOwnedbyAliCloud(ctx, shoot, imageId, region)
+	return true, nil
+}
+func (s *shootMutator) isOwnedbyAliCloud(ctx context.Context, shoot *corev1beta1.Shoot, imageId string, region string) (bool, error) {
+
+	var (
+		secretBinding    = &corev1beta1.SecretBinding{}
+		secretBindingKey = kutil.Key(shoot.Namespace, shoot.Spec.SecretBindingName)
+	)
+	if err := kutil.LookupObject(ctx, s.virtualGardenclient, s.apiReader, secretBindingKey, secretBinding); err != nil {
+		return false, err
+	}
+
+	var (
+		secret    = &corev1.Secret{}
+		secretRef = secretBinding.SecretRef.Name
+		secretKey = kutil.Key(secretBinding.SecretRef.Namespace, secretRef)
+	)
+	if err := s.apiReader.Get(ctx, secretKey, secret); err != nil {
+		return false, err
+	}
+	accessKeyID, ok := secret.Data[alicloud.AccessKeyID]
+	if !ok {
+		return false, fmt.Errorf("missing %q field in secret %s", alicloud.AccessKeyID, secretRef)
+	}
+	accessKeySecret, ok := secret.Data[alicloud.AccessKeySecret]
+	if !ok {
+		return false, fmt.Errorf("missing %q field in secret %s", alicloud.AccessKeySecret, secretRef)
+	}
+	shootECSClient, err := s.alicloudClientFactory.NewECSClient(region, string(accessKeyID), string(accessKeySecret))
+	if err != nil {
+		return false, err
+	}
+	if exist, err := shootECSClient.CheckIfImageExists(ctx, imageId); err != nil {
+		return false, err
+	} else if exist {
+		if ownedByAliCloud, err := shootECSClient.CheckIfImageOwnedByAliCloud(imageId); err != nil {
+			return false, err
+		} else if ownedByAliCloud {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+func (s *shootMutator) getImageId(ctx context.Context, imageName string, imageVersion *string, imageRegion string, cloudProfileName string) (string, error) {
+	var (
+		cloudProfile    = &corev1beta1.CloudProfile{}
+		cloudProfileKey = kutil.Key(cloudProfileName)
+	)
+	imageId := ""
+	if err := kutil.LookupObject(ctx, s.virtualGardenclient, s.apiReader, cloudProfileKey, cloudProfile); err != nil {
+		return imageId, err
+	}
+	cloudProfileConfig, err := s.getCloudProfileConfig(cloudProfile)
+	if err != nil {
+		return imageId, err
+	}
+	imageId, err = helper.FindImageForRegionFromCloudProfile(cloudProfileConfig, imageName, *imageVersion, imageRegion)
+	return imageId, err
+}
+func (s *shootMutator) getCloudProfileConfig(cloudProfile *corev1beta1.CloudProfile) (*api.CloudProfileConfig, error) {
+	var cloudProfileConfig *api.CloudProfileConfig = &api.CloudProfileConfig{}
+	if _, _, err := s.decoder.Decode(cloudProfile.Spec.ProviderConfig.Raw, nil, cloudProfileConfig); err != nil {
+		return nil, fmt.Errorf("could not decode providerConfig of cloudProfile for '%s': %w", kutil.ObjectName(cloudProfile), err)
+	}
 
-func (s *shootMutator) mutateShootUpdate(oldShoot, shoot *corev1beta1.Shoot) error {
+	return cloudProfileConfig, nil
+}
+func (s *shootMutator) mutateShootUpdate(ctx context.Context, oldShoot, shoot *corev1beta1.Shoot) error {
+	if !equality.Semantic.DeepEqual(oldShoot.Spec, shoot.Spec) {
```

Comment:
dup with line 191

Reply:
Hi Jerry,
I did this on purpose. After the first mutate action, the shoot spec will be changed and then it might be already same with old spec. then we can avoid unnecessary call later.

Reply:
OK, it is fine though `mutateForEncryptedSystemDiskChange` has no side effect so far. Would you please add some comment here to make readers understand why you specially check it twice ?:)